File: gol/csharp/src/main/csharp/GameOfLife.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace csharp.main.csharp
{
    /// <summary>
    ///     Implementation for the Advent Coding Contest 2017 in C#
    /// </summary>
    /// <author>
    ///     Marc Rousavy (marcrousavy@hotmail.com)
    /// </author>
    public class GameOfLife {
        /// <summary>
        ///     The constructor of GameOfLife initializes the pattern and all the private attributes
        /// </summary>
        /// <param name="initPattern">
        ///     should obey this definition for a pattern <code>"3;3;0;110001110$"</code>, where the information is
        ///     defined as <code>"width;height;generation;[row][row][row]$"</code> and the dollar($) at the end marks
        ///     the End-Of-Line.
        /// </param>
        public GameOfLife(string initPattern) {
            string[] info = initPattern.Split(";");
            Width = int.Parse(info[0]);
            Height = int.Parse(info[1]);
            Generation = int.Parse(info[2]);
            Pattern = info[3].Substring(0, info[3].Length - 1);
            RuleSet = new HashSet<string>();
        }

        private int Width { get; }

        private int Height { get; }

        private int Generation { get; set; }

        private string Pattern { get; }

        private HashSet<string> RuleSet { get; }

        /// <summary>
        ///     Returns a pattern as a String for the given generation number.
        /// </summary>
        /// <param name="generation">
        ///     The generation for which the pattern should be returned.
        /// </param>
        /// <returns>
        ///     should obey this definition for a pattern "3;3;0;110001110$", where the information is defined as
        ///     <code>"width;height;generation;[row][row][row]$"</code> and the dollar ($) at the end marks the End-Of-Line.
        /// </returns>
        public string GetPattern(int generation) {
            // TODO: Here you should evaluate the pattern


            var ret = new StringBuilder();
            ret.Append(Width + ";");
            ret.Append(Height + ";");
            ret.Append(Generation + ";");
            ret.Append(Pattern + "$\n");
            return ret.ToString();
        }

        /// <summary>
        ///     Returns the pattern of the next generation.
        /// </summary>
        /// <returns>
        ///     should obey this definition for a pattern "3;3;0;110001110$", where
        ///     the information is defined as <code>"width;height;generation;[row][row][row]$"</code>
        ///     and the dollar ($) at the end marks the End-Of-Line.
        /// </returns>
        public string GetNextPattern() {
            return GetPattern(++Generation);
        }

        /// <summary>
        ///     Returns the pattern of the previous generation.
        /// </summary>
        /// <returns>
        ///     should obey this definition for a pattern "3;3;0;110001110$", where
        ///     the information is defined as <code>"width;height;generation;[row][row][row]$"</code>
        ///     and the dollar ($) at the end marks the End-Of-Line.
        /// </returns>
        public string GetPreviousPattern() {
            return GetPattern(--Generation);
        }

        /// <summary>
        ///     Adding one rule to the Game-Of-Life Engine rule-set.
        /// </summary>
        /// <param name="descriptionId">
        ///     A String which defines a Game-Of-Life rule.
        /// </param>
        public void AddRule(string descriptionId) {
            RuleSet.Add(descriptionId);
        }

        /// <summary>
        ///     Adding rules to the Game-Of-Life Engine rule-set.
        /// </summary>
        /// <param name="descriptionIdList">
        ///     A List of Strings with Game-Of-Life rules.
        /// </param>
        public void AddRules(List<string> descriptionIdList) {
            descriptionIdList.ForEach(AddRule);
        }

        /// <summary>
        ///     Removing one rule from the Game-Of-Life Engine rule-set.
        /// </summary>
        /// <param name="descriptionId">
        ///     A String which defines a Game-Of-Life rule.
        /// </param>
        public void RemoveRule(string descriptionId) {
            RuleSet.Remove(descriptionId);
        }

        /// <summary>
        ///     Removing rules from the Game-Of-Life Engine rule-set.
        /// </summary>
        /// <param name="descriptionIdList">
        ///     A List of Strings with Game-Of-Life rules.
        /// </param>
        public void RemoveRules(List<string> descriptionIdList) {
            descriptionIdList.ForEach(RemoveRule);
        }

        /// <summary>
        ///     Exporting the patterns to a file. Example:
        ///     * <code>2;2;0;1011$ 2;2;1;0000$</code>
        ///     * <code>2;2;2;0000$ 2;2;3;0000$ 2;2;4;0000$</code>
        /// </summary>
        /// <param name="start">
        ///     First generation number.
        /// </param>
        /// <param name="end">
        ///     Last generation number.
        /// </param>
        /// <param name="filename">
        ///     Name of File.
        /// </param>
        public void ExportGenerations(int start, int end, string filename) {
            var output = new StringBuilder();

            if (start <= end)
                for (int i = start; i <= end; i++)
                    output.Append(GetPattern(i));
            else
                for (int i = end; i >= start; i--)
                    output.Append(GetPattern(i));

            try {
                File.WriteAllText(filename, output.ToString());
            } catch (Exception e) {
                Console.WriteLine(e);
            }
        }
    }
}

File: gol/java/src/main/java/GameOfLife.java
package gol;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;
import java.util.Set;

/**
 * Implementation for the Advent Coding Contest 2017
 * 
 * @author Michael Borko (mborko@tgm.ac.at)
 *
 */
public class GameOfLife {

	private Integer width = 0;
	private Integer height = 0;
	private Integer generation = 0;
	private String pattern;
	private Set<String> ruleSet;

	/**
	 * The constructor of GameOfLife initializes the pattern and all the private
	 * attributes.
	 * 
	 * @param initPattern
	 *            should obey this definition for a pattern "3;3;0;110001110$",
	 *            where the information is defined as
	 *            "width;height;generation;[row][row][row]$" and the dollar ($) at
	 *            the end marks the End-Of-Line.
	 */
	public GameOfLife(String initPattern) {
		String[] info = initPattern.split(";");
		this.width = Integer.parseInt(info[0]);
		this.height = Integer.parseInt(info[1]);
		this.generation = Integer.parseInt(info[2]);
		this.pattern = info[3].substring(0, info[3].length() - 1);
	}

	/**
	 * Returns a pattern as a String for the given generation number.
	 * 
	 * @param generation
	 *            The generation for which the pattern should be returned.
	 * @return should obey this definition for a pattern "3;3;0;110001110$", where
	 *         the information is defined as
	 *         "width;height;generation;[row][row][row]$" and the dollar ($) at the
	 *         end marks the End-Of-Line.
	 */
	public String getPattern(Integer generation) {

		// Here you should evaluate the pattern

		StringBuffer ret = new StringBuffer();
		ret.append(width + ";");
		ret.append(height + ";");
		ret.append(this.generation + ";");
		ret.append(pattern + "$\n");
		return ret.toString();
	}

	/**
	 * Returns the pattern of the next generation.
	 * 
	 * @return should obey this definition for a pattern "3;3;0;110001110$", where
	 *         the information is defined as
	 *         "width;height;generation;[row][row][row]$" and the dollar ($) at the
	 *         end marks the End-Of-Line.
	 */
	public String getNextPattern() {
		return getPattern(++generation);
	}

	/**
	 * Returns the pattern of the previous generation.
	 * 
	 * @return should obey this definition for a pattern "3;3;0;110001110$", where
	 *         the information is defined as
	 *         "width;height;generation;[row][row][row]$" and the dollar ($) at the
	 *         end marks the End-Of-Line.
	 */
	public String getPreviousPattern() {
		return getPattern(--generation);
	}

	/**
	 * Adding one rule to the Game-Of-Life Engine rule-set.
	 * 
	 * @param descriptionID
	 *            A String which defines a Game-Of-Life rule.
	 */
	public void addRule(String descriptionID) {
		ruleSet.add(descriptionID);
	}

	/**
	 * Adding rules to the Game-Of-Life Engine rule-set.
	 * 
	 * @param descriptionIDList
	 *            A List of Strings with Game-Of-Life rules.
	 */
	public void addRules(List<String> descriptionIDList) {
		for (String id : descriptionIDList)
			addRule(id);
	}

	/**
	 * Removing one rule from the Game-Of-Life Engine rule-set.
	 * 
	 * @param descriptionID
	 *            A String which defines a Game-Of-Life rule.
	 */
	public void removeRule(String descriptionID) {
		ruleSet.remove(descriptionID);
	}

	/**
	 * Removing rules from the Game-Of-Life Engine rule-set.
	 * 
	 * @param descriptionIDList
	 *            A List of Strings with Game-Of-Life rules.
	 */
	public void removeRules(List<String> descriptionIDList) {
		for (String id : descriptionIDList)
			removeRule(id);
	}

	/**
	 * Exporting the patterns to a file. Example: 2;2;0;1011$ 2;2;1;0000$
	 * 2;2;2;0000$ 2;2;3;0000$ 2;2;4;0000$
	 * 
	 * @param start
	 *            First generation number.
	 * @param end
	 *            Last generation number.
	 * @param filename
	 *            Name of File.
	 */
	public void exportGenerations(Integer start, Integer end, String filename) {
		StringBuffer output = new StringBuffer("");
		if (start <= end)
			for (int i = start; i <= end; i++)
				output.append(getPattern(i));
		else
			for (int i = end; i >= start; i--)
				output.append(getPattern(i));

		try {
			BufferedWriter writer;
			writer = new BufferedWriter(new FileWriter(new File(filename)));
			writer.write(output.toString());
			writer.flush();
			writer.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	/**
	 * Check your functionality with this main method.
	 * 
	 * @param args
	 *            Define as first parameter the pattern (e.g.
	 *            "4;4;0;1011101010101001$"), second the filename, third the start
	 *            generation and as last parameter the end generation.
	 *            E.g. java gol.GameOfLife 4;4;0;1011101010101001$ out.csv 0 100
	 */
	public static void main(String[] args) {
		GameOfLife gol;
		if (args.length != 4) {
			gol = new GameOfLife("4;4;0;1011101010101001$");
			gol.exportGenerations(0, 5, "output.csv");
		} else if (args.length == 4) {
			try {
				gol = new GameOfLife(args[0]);
				gol.exportGenerations(Integer.parseInt(args[2]), Integer.parseInt(args[3]), args[1]);
			} catch (NumberFormatException e) {
				e.printStackTrace();
			}
		}
	}
}
